lib/Optimizer/Transforms/ExpandControlVeqs.cpp

```diff
@@ -0,0 +1,131 @@
+
+namespace {
+template <typename OP>
+class ExpandPat : public OpRewritePattern<OP> {
+
+public:
+  using OpRewritePattern<OP>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(OP op,
+                                PatternRewriter &rewriter) const override {
+    // Unfortunately, we may grow the # of controls as a result of this rewrite,
+    // so we must track a new list of controls and reconstruct the operands
+    // instead of replacing the controls in place.
+    SmallVector<Value> newControls;
+    bool update = false;
+
+    // Search through the controls for veqs with known sizes
+    for (auto [index, control] : llvm::enumerate(op.getControls())) {
+      if (auto veq = dyn_cast<quake::VeqType>(control.getType())) {
+        if (!veq.hasSpecifiedSize())
+          return failure();
+
+        // For each of the qubits in the veq, create an extraction instruction
+        // The result of the extraction will be a new control
+        // The veq is not added the newControls, so it will be dropped
+        for (size_t i = 0; i < veq.getSize(); ++i) {
+          auto ext =
+              rewriter.create<quake::ExtractRefOp>(op.getLoc(), control, i);
+          newControls.push_back(ext);
+          update = true;
+        }
+      } else {
+        newControls.push_back(control);
+      }
+    }
+
+    if (!update)
+      return failure();
+
+    // Reconstruct the operation with the new controls
+    auto segmentSizes = rewriter.getDenseI32ArrayAttr(
+        {static_cast<int32_t>(op.getParameters().size()),
+         static_cast<int32_t>(newControls.size()),
+         static_cast<int32_t>(op.getTargets().size())});
+
+    auto newOp = rewriter.replaceOpWithNewOp<OP>(
+        op, op.getIsAdj(), op.getParameters(), newControls, op.getTargets(),
+        op.getNegatedQubitControlsAttr());
+
+    newOp->setAttr("operand_segment_sizes", segmentSizes);
+
+    return success();
+  }
+};
+
+struct ExpandControlVeqsPass
+    : public cudaq::opt::impl::ExpandControlVeqsBase<ExpandControlVeqsPass> {
+  using ExpandControlVeqsBase::ExpandControlVeqsBase;
+
+private:
+  template <typename OP>
+  static bool checkLegal(OP op) {
+    for (auto control : op.getControls()) {
+      // Valid ops have no control veqs with a known size
+      if (auto veq = dyn_cast<quake::VeqType>(control.getType()))
+        if (veq.hasSpecifiedSize())
+          return false;
+    }
+
+    return true;
+  }
+
+public:
+  void runOnOperation() override {
+    auto *ctx = &getContext();
+    func::FuncOp func = getOperation();
+    RewritePatternSet patterns(ctx);
+    patterns.insert<
+        ExpandPat<quake::HOp>, ExpandPat<quake::PhasedRxOp>,
+        ExpandPat<quake::R1Op>, ExpandPat<quake::RxOp>, ExpandPat<quake::RyOp>,
+        ExpandPat<quake::RzOp>, ExpandPat<quake::SOp>, ExpandPat<quake::SwapOp>,
+        ExpandPat<quake::TOp>, ExpandPat<quake::U2Op>, ExpandPat<quake::U3Op>,
+        ExpandPat<quake::XOp>, ExpandPat<quake::YOp>, ExpandPat<quake::ZOp>>(
```

Comment:
Is it possible to shorten these lists by using the `OperatorInterface` interface? (I believe these all support that interface.)